sympy/core/function.py

```diff
@@ -1930,26 +1971,42 @@ def __call__(self, *args):
             ## XXX does this apply to Lambda? If not, remove this comment.
             temp = ('%(name)s takes exactly %(args)s '
                    'argument%(plural)s (%(given)s given)')
-            raise TypeError(temp % {
+            raise BadArgumentsError(temp % {
                 'name': self,
                 'args': list(self.nargs)[0],
                 'plural': 's'*(list(self.nargs)[0] != 1),
                 'given': n})
-        return self.expr.xreplace(dict(list(zip(self.variables, args))))
+
+        d = self._match_signature(self.signature, args)
+
+        return self.expr.xreplace(d)
+
+    def _match_signature(self, sig, args):
+
+        symargmap = {}
+
+        def rmatch(pars, args):
+            for par, arg in zip(pars, args):
+                if par.is_Symbol:
+                    symargmap[par] = arg
+                elif isinstance(par, Tuple):
+                    if not isinstance(arg, (tuple, Tuple)) or len(args) != len(pars):
+                        raise BadArgumentsError("Can't match %s and %s" % (args, pars))
+                    rmatch(par, arg)
+
+        rmatch(sig, args)
+
+        return symargmap
 
     def __eq__(self, other):
         if not isinstance(other, Lambda):
             return False
         if self.nargs != other.nargs:
             return False
 
-        selfexpr = self.args[1]
-        otherexpr = other.args[1]
-        otherexpr = otherexpr.xreplace(dict(list(zip(other.args[0], self.args[0]))))
-        return selfexpr == otherexpr
+        canonical = lambda f: f.xreplace(f.canonical_variables)
 
-    def __ne__(self, other):
-        return not(self == other)
+        return canonical(self).args == canonical(other).args
 
     def __hash__(self):
         return super(Lambda, self).__hash__()
```

Comment:
I don't understand why this method is necessary. The default for calling `A.__hash__` where class `A` does not define `__hash__` should be to call the superclass' hash method. Removing this function made `Lambda` unhashable on Python 3 though.

Reply:
It turns out that in Python 3 a class that overrides `__eq__` cannot automatically inherit `__hash__` since the two must be overridden together. That does not quite apply here since `_hashable_content` is overridden which is the main part of `Basic/Expr.__hash__`. This set up does lead to hash problems in master though (#17498).